typed-racket-lib/typed-racket/types/subtype.rkt

```diff
@@ -384,397 +396,560 @@
+                  (with-updated-seen A
+                    (subtype* A parent1 t2))))]
+          [else (continue)])]
+     [_ (continue)])]
+  [(case: StructType (StructType: t1))
+   (match t2
+     [(StructTypeTop:) A]
+     [_ (continue)])]
+  [(case: Syntax (Syntax: elem1))
+   (match t2
+     [(Syntax: elem2) (subtype* A elem1 elem2)]
+     [_ (continue)])]
+  [(case: ThreadCell (ThreadCell: elem1))
+   (match t2
+     [(? ThreadCellTop?) A]
+     [(ThreadCell: elem2) (type-equiv? A elem1 elem2)]
+     [_ (continue)])]
+  [(case: Union (Union: elems1))
+   (cond
+     [(cache-ref union-sub-cache t1 t2)
+      => (λ (b) (and (unbox b) A))]
+     [else
+      (define result
+        (match t2
+          [(Union: elems2)
+           (for/fold ([A A])
+                     ([elem1 (in-hset elems1)]
+                      #:break (not A))
+             (if (hset-member? elems2 elem1)
+                 A
+                 (subtype* A elem1 t2)))]
+          [_ (for/fold ([A A])
+                       ([elem1 (in-hset elems1)]
+                        #:break (not A))
+               (subtype* A elem1 t2))]))
+      (when (null? A)
+        (cache-set! union-sub-cache t1 t2 (and result #t)))
+      result])]
+  ;; For Unit types invoke-types are covariant
+  ;; imports and init-depends are covariant in that importing fewer
+  ;; signatures results in a subtype
+  ;; exports conversely are contravariant, subtypes export more signatures
+  [(case: Unit (Unit: imports1 exports1 init-depends1 t1))
+   (match t2
+     [(? UnitTop?) A]
+     [(Unit: imports2 exports2 init-depends2 t2)
+      (and (check-sub-signatures? imports2 imports1)
+           (check-sub-signatures? exports1 exports2)
+           (check-sub-signatures? init-depends2 init-depends1)
+           (subval* A t1 t2))]
+     [_ (continue)])]
+  [(case: Value (Value: val1))
+   (match t2
+     [(Base: _ _ pred _) (and (pred val1) A)]
+     [(Sequence: (list seq-t))
+      (cond
+        [(null? val1) A]
+        [(exact-nonnegative-integer? val1)
+         (define possibilities
+           (list
+            (list byte? -Byte)
+            (list portable-index? -Index)
+            (list portable-fixnum? -NonNegFixnum)
+            (list values -Nat)))
+         (define type
+           (for/or ([pred-type (in-list possibilities)])
+             (match pred-type
+               [(list pred? type)
+                (and (pred? val1) type)])))
+         (subtype* A type seq-t)]
+        [else #f])]
+     [(or (? Struct? s1) (NameStruct: s1))
+      #:when (not (struct? val1))
+      #f]
+     [_ (continue)])]
+  [(case: Vector (Vector: elem1))
+   (match t2
+     [(? VectorTop?) A]
+     [(Vector: elem2) (type-equiv? A elem1 elem2)]
+     [(Sequence: (list seq-t)) (subtype* A elem1 seq-t)]
+     [_ (continue)])]
+  [(case: Weak-Box (Weak-Box: elem1))
+   (match t2
+     [(? Weak-BoxTop?) A]
+     [(Weak-Box: elem2) (type-equiv? A elem1 elem2)]
+     [_ (continue)])]
+  [else: (continue)])
```

Comment:
The control flow here is slightly annoying.

Currently the switch will return #t or #f if the subtyping can be determined, or it calls the continuation `continue` to finish checking the remaining subtyping possibilities (which are based on the rhs argument).

I could return a sentinel value and check for the "we haven't figured out yes or no" case... or we can do this continuation calling.